agent/lib_monolog.c

```diff
@@ -306,6 +309,150 @@ static char* nr_monolog_get_context(const size_t argc,
+ *           Other zval types are also ignored.
+ */
+nrobj_t* nr_monolog_context_data_zval_to_attribute_obj(
+    const zval* z TSRMLS_DC) {
+  nrobj_t* retobj = NULL;
+
+  if (NULL == z) {
+    return NULL;
+  }
+
+  nr_php_zval_unwrap(z);
+
+  switch (Z_TYPE_P(z)) {
+    case IS_NULL:
+      retobj = NULL;
+      break;
+
+    case IS_LONG:
+      retobj = nro_new_long((long)Z_LVAL_P(z));
+      break;
+
+    case IS_DOUBLE:
+      retobj = nro_new_double(Z_DVAL_P(z));
+      break;
+
+    case IS_TRUE:
+      retobj = nro_new_boolean(true);
+      break;
+
+    case IS_FALSE:
+      retobj = nro_new_boolean(false);
+      break;
+
+    case IS_STRING:
+      if (!nr_php_is_zval_valid_string(z)) {
+        retobj = NULL;
+      } else {
+        retobj = nro_new_string(Z_STRVAL_P(z));
+      }
+      break;
+
+    default:
+      /* any other type conversion to attribute not supported */
+      retobj = NULL;
+      break;
+  }
+
+  return retobj;
+}
+
+/*
+ * Purpose : Extract $context argument of Monolog\Logger::addRecord to
+ * attributes
+ *
+ * Params  : # of Monolog\Logger::addRecord arguments, and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : zval* for context array on success (must be freed by caller)
+ *           NULL otherwise
+ *
+ */
+static zval* nr_monolog_extract_context_data(const size_t argc,
+                                             NR_EXECUTE_PROTO TSRMLS_DC) {
+  zval* context_arg;
+
+  if (3 > argc) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: $context not available", __func__);
+    goto return_context;
+  }
+
+  context_arg = nr_php_arg_get(3, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+
+  if (NULL == context_arg) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: $context not defined, unable to get log context",
+                     __func__);
+    goto return_context;
+  } else if (!nr_php_is_zval_valid_array(context_arg)) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: expected $context be a valid array, got type %d",
+                     __func__, Z_TYPE_P(context_arg));
+    goto return_context;
+  }
+
+  if (0 == nr_php_zend_hash_num_elements(Z_ARRVAL_P(context_arg))) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: $context has no elements", __func__);
+    goto return_context;
+  }
+
+return_context:
+  return context_arg;
+}
+
+/*
+ * Purpose : Convert $context array of Monolog\Logger::addRecord to
+ * attributes
+ *
+ * Params  : zval* for context array from Monolog
+ *
+ * Returns : json representation on success
+ *           NULL otherwise
+ *
+ */
+nr_attributes_t* nr_monolog_convert_context_data_to_attributes(
+    zval* context_data TSRMLS_DC) {
+  zend_string* key;
+  zval* val;
+
+  nr_attributes_t* attributes = NULL;
+
+  if (NULL == context_data || !nr_php_is_zval_valid_array(context_data)) {
+    return NULL;
+  }
+
+  attributes = nr_attributes_create(NRPRG(txn)->attribute_config);
+
+  ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARR_P(context_data), key, val) {
```

Comment:
Will this work if `$context` is just an array and not a hashmap and the keys are numeric? Can `$context` be just an array with numeric keys? I'd suggest looking at and potentially re-using some pieces of `nr_monolog_fmt_context` (which uses more generic `nr_php_zend_hash_zval_apply`), `nr_monolog_fmt_context_item`, `nr_monolog_fmt_context_key`, `nr_monolog_fmt_context_value`.

Reply:
Monolog sends a hash currently so I don't know if this is a concern for code that is focused on the Monolog library.

Reply:
The current agent attribute implementation only accepts strings as attribute keys so any elements using an integer key are ignored.

Reply:
Fair enough. However, I just wanted to point out that this is valid code:
```
$logger->alert("context is array", ["foobar"]);
```
which yields following output when using `LineFormatter("%level_name%: %message% %context%\n");`:
```
ALERT: context is array ["foobar"]
```

Reply:
I see there is a test for that - test_monolog_context_simple.php